[solax] Initial contribution (#14880)
openhab/openhab-addons: bundles/org.openhab.binding.solax/README.md

# Solax Binding

This is a binding for Solax solar power inverters.

Solax Wi-Fi module with direct connection via HTTP is supported.
Wi-Fi module firmware version 3.x+ is required.
Please note that earlier firmware releases do not support direct connection, therefore the binding will not work in its current state.

The binding retrieves a structured data from the inverter's Wi-Fi module, parses it and pushes it into the inverter Thing where each channel represents a specific information (inverter output power, voltage, PV1 power, etc.)

In case the parsed information that comes with the binding out of the box differs, the raw data channel can be used with a combination of JSON Path transformation to map the proper values to the necessary items.

## Supported Things

> | Thing                | Thing Type | Description                                                                         |
> |----------------------|------------|-------------------------------------------------------------------------------------|
> | localConnectInverter | Thing      | This is model representation of inverter with all the data available as a channels  |

Reviewer: ```suggestion
| Thing                  | Thing Type | Description                                                                         |
|------------------------|------------|-------------------------------------------------------------------------------------|
| local-connect-inverter | Thing      | This is model representation of inverter with all the data available as a channels  |
```

|--------------------------|----------------------------|---------------------------------------------------------------------------------------------------------------------------------------------|
| last-update-time         | DateTime                   | Last time when a call has been made to the inverter                                                                                         |
| raw-data                 | String                     | The raw data retrieved from inverter in JSON format. (Usable for channels not implemented. Can be consumed with the JSONpath transformation |

### Properties

| Property          | Description                               |
|-------------------|-------------------------------------------|
| serialNumber      | The serial number of the Wi-Fi module     |
| inverterType      | Inverter Type (for example X1_HYBRID_G4)  |

## Full Example

Here are some file based examples.

### Thing Configuration

```java
// The local connect inverter thing 

> Thing solax:localConnectInverter:localInverter  [ refreshInterval=10, password="<SERIAL NUMBER OF THE WIFI MODULE>", hostname="<local IP/hostname in the network>" ] 

Reviewer: ```suggestion
Thing solax:local-connect-inverter:localInverter  [ refreshInterval=10, password="<SERIAL NUMBER OF THE WIFI MODULE>", hostname="<local IP/hostname in the network>" ] 
```

# Solax Binding

This is a binding that provides data of Solax solar power inverters.

Currently it supports only a Solax Wi-Fi module with direct connection via HTTP (Wi-Fi module firmware version 3.x+ is required).
Please note that earlier firmware releases do not support direct connection, therefore the binding will not work in its current state.

The binding retrieves a structured data from the inverter's Wi-Fi module, parses it and pushes it into the inverter Thing where each channel represents a specific information (inverter output power, voltage, PV1 power, etc.)

In case the parsed information that comes with the binding out of the box differs, the raw data channel can be used with a combination of JSON Path transformation to map the proper values to the necessary items.

> 

Author: Why?

> 

Reviewer: IIRC there were two linefeeds after each other.

# Solax Binding

This is a binding that provides data of Solax solar power inverters.

Currently it supports only a Solax Wi-Fi module with direct connection via HTTP (Wi-Fi module firmware version 3.x+ is required).

> Please note that earlier firmware releases do not support direct connection, therefore the binding will not work in it's current state.

Reviewer: ```suggestion
Please note that earlier firmware releases do not support direct connection, therefore the binding will not work in its current state.
```

| inverter          | Thing      | This is model representation of inverter with all the data available as a channels (Ex. inverter output power, voltage, PV1 power, etc)     |

## Thing Configuration

### Local Connect Inverter Configuration

| Parameter         | Description                                                                                                                                        |
|-------------------|----------------------------------------------------------------------------------------------------------------------------------------------------|
| refreshInterval   | Value is in seconds. Defines the refresh interval when the binding polls from the inverter's Wi-Fi module. Optional parameter. Default 10 seconds. |
| password          | Password for accessing the Wi-Fi module (the serial number of the wifi). Mandatory parameter.                                                      |
| hostname          | IP address or hostname of your Wi-Fi module. If hostname is used must be resolvable by OpenHAB. Mandatory parameter.                               |

> The bridge does not have any channels. 
> It is used only to connect to the module and retrieve the data. 
> The data is shown in the Inverter thing

Reviewer: Now simplified:
```suggestion
```

| feed-in-power            | Number:Power               | The power to / from grid (negative means power is pulled from the grid and vice-versa) [W]     |

### General channels

| Channel                  | Type                       | Description                                                                                                                                 |
|--------------------------|----------------------------|---------------------------------------------------------------------------------------------------------------------------------------------|
| last-update-time         | DateTime                   | Last time when a call has been made to the inverter                                                                                         |
| raw-data                 | String                     | The raw data retrieved from inverter in JSON format. (Usable for channels not implemented. Can be consumed with the JSONpath transformation |

### Properties

| Property          | Description                               |
|-------------------|-------------------------------------------|

> | serialNumber      | The Serial Number of the Wi-Fi module     |

Reviewer: ```suggestion
| serialNumber      | The serial number of the Wi-Fi module     |
```

| inverterType      | Inverter Type (for example X1_HYBRID_G4)  |

## Full Example

Here are some file based examples.

### Thing Configuration

```java
// The local connect inverter thing 
Thing solax:localConnectInverter:localInverter  [ refresh=10, password="<SERIAL NUMBER OF THE WIFI MODULE>", hostname="<local IP/hostname in the network>" ] 
```

### Item Configuration

```java
Group gSolaxInverter "Solax Inverter" <energy> (boilerRoom)
Group solarPanels "Solar panels" <energy> (gSolaxInverter)

> String solaxSerial "Serial Number [%s]" <energy> (gsolax_inverter) { channel="solax:localConnectInverter:localInverter:serialNumber" }

Author: So the property even though updated via the thing handler cannot act as a channel and be supplied to an item? I was wondering why it does not work in my own sitemap...

> String solaxSerial "Serial Number [%s]" <energy> (gsolax_inverter) { channel="solax:localConnectInverter:localInverter:serialNumber" }

Reviewer: No, channels and properties are different. Some confusion might occur from the fact that discovery uses properties to populate configuration parameters for discovered things.

# Solax Binding

This is a binding for a Solax Wi-Fi module that supports connection directly via HTTP (Wi-Fi module firmware version 3.x+ is required).
Please note that earlier firmware releases do not support direct connection, therefore the binding will not work in it's current state.

Currently the Solax cloud services provide an update every 5 minutes (sometimes even more rarely) and they also happen to be down sometimes, which makes it hard to automate the decision making by openHAB if we have more complex rules.

The binding retrieves a structured data from the Wi-Fi module, parses it and pushes it into the inverter Thing where each channel represents a specific information (inverter output power, voltage, PV1 power, etc)

## Supported Things

| Thing             | Thing Type | Description                                                                                                                                 |
|-------------------|------------|---------------------------------------------------------------------------------------------------------------------------------------------|

> | localConnect      | Bridge     | The bridge is used to communicate directly to the wifi module of the inverter.                                                              |

Reviewer: ```suggestion
| localConnect      | Bridge     | The bridge is used to communicate directly with the Wi-Fi module of the inverter.                                                           |
```

|-------------------|----------------------------------------------------------------------------------------------------------------------------------------------------|
| refresh           | Value is in seconds. Defines the refresh interval when the binding polls from the inverter's Wi-Fi module. Optional parameter. Default 10 seconds. |
| password          | Password for accessing the Wi-Fi module (the serial number of the wifi). Mandatory parameter.                                                      |
| hostname          | IP address or hostname of your Wi-Fi module. If hostname is used must be resolvable by OpenHAB. Mandatory parameter.                               |

The bridge does not have any channels. It is used only to connect to the module and retrieve the data. The data is shown in the Inverter thing

### Inverter Thing Configuration

### Inverter Output Channels

| Channel                  | Type                       | Description                                      |
|--------------------------|----------------------------|--------------------------------------------------|

> | inverterOutputPower      | system.electric-power      | The output power of the inverter [W]             |

Reviewer: Channel names need to be adapted after renaming.

| inverterCurrent          | system.electric-current    | The output current of the inverter [A]           |
| inverterVoltage          | system.electric-voltage    | The output voltage of the inverter [V]           |
| inverterFrequency        | Number:ElectricPotential   | The frequency of the output voltage [Hz]         |

### Photovoltaic Panels Production Channels

| Channel                  | Type                       | Description                            |
|--------------------------|----------------------------|----------------------------------------|

> | pv1Voltage      | system.electric-voltage    | The voltage of PV1 string [V]                   |

Reviewer: Channel names need to be adapted after renaming.

### General channels

| Channel                  | Type                       | Description                                                                                                                                 |
|--------------------------|----------------------------|---------------------------------------------------------------------------------------------------------------------------------------------|
| lastUpdateTime           | DateTime                   | Last time when a call has been made to the inverter                                                                                         |
| rawData                  | String                     | The raw data retrieved from inverter in JSON format. (Usable for channels not implemented. Can be consumed with the JSONpath transformation |

### Properties

| Property          | Description                               |
|-------------------|-------------------------------------------|

> | serialWifi        | The Wi-Fi module's serial number          |

Reviewer: I believe this is now:
```suggestion
| serialNumber      | The Wi-Fi module's serial number          |
```

# Solax Binding

> This is a OpenHAB binding for a Solax Wi-fi modules that have the option to be connected directly via HTTP (firmware version 3.x+).

Reviewer: I meant to mention the type of device supported, for example: "This is a binding for Solax solar inverters".

Please note that earlier firmware releases do not support direct connection.

Currently the Solax cloud services provide an update on every 5 minutes (sometimes even more) and they also happen to be down sometimes, which makes it hard to automate the decision making by OpenHAB if we have more complex rules.

For example: I want to charge my car if the battery of my inverter is 90%+ and my overall consumption is only from the solar power.
The binding retrieves structured data from the wifi module, parses it and pushes it into the inverter thing where each channel represents a specific information (inverter output power, voltage, PV1 power, etc)

Hostname/IP address of the wi-fi module (can be any IP address from internal DHCP wi-fi network) [String]

> The serial number of the Wi-fi module is the default password (not sure if it can be changed) [String]

Author: removed them. Not needed anymore

Time interval to poll the data from the module [number in seconds]

> Please give it a try and provide me a feedback if it works for you. I will update the releases as soon as I implement some of the backlog.

Reviewer: This documentation is probably not the best place for this comment. Perhaps some kind of announcement in the forum would be better?

> ## Supported Things

Reviewer: ```suggestion

## Supported Things
```

| localConnect      | Bridge     | The bridge is used to communicate directly to the wifi module of the inverter.|
| inverter          | Thing      | This is model representation of inverter with all the data available as a channels (Ex. inverter output power, voltage, PV1 power, etc) |

## Discovery

Discovery is available. Once the localConnect bridge is configured with the necessary parameters, it can discover automatically the inverter with all it's channels as a separate thing.

## Thing Configuration

### Local Connect Bridge parameters

| Parameter         | Description                            |
|-------------------|----------------------------------------|
| refresh           | Value is in seconds. Defines the refresh interval when the binding polls from the inverter's wifi module. Optional parameter. Default 10 seconds. |
| password          | Password for accessing the Wifi module (the serial number of the wifi). Mandatory parameter.  |
| hostname          | IP address or hostname of your Wifi module. If hostname is used must be resolvable by OpenHAB. Mandatory parameter.  |

> The bridge does not have any channels. It is used only to connect to the module and retrieve the data. The data is shown in the Inverter thing

Reviewer: If you put each sentence in a separate line, reading the diffs would be easier.
```suggestion
The bridge does not have any channels.
It is used only to connect to the module and retrieve the data.
The data is shown in the Inverter Thing.
```

### Inverter Thing Configuration

### Inverter Output Channels

| Channel                  | Type                       | Description                                                                               |
|--------------------------|----------------------------|-------------------------------------------------------------------------------------------|

> | inverterOutputPower      | system.electric-power      | The output power of the inverter [W]           |

Reviewer: Should be item type, not channel type:
```suggestion
| inverterOutputPower      | Number:Power               | The output power of the inverter [W]           |
```

Please fix all.

Author: Can you give me a hint where to see these?

Reviewer: You should be able to see them in the UI, but you could also cheat by looking here 😄
https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/DefaultSystemChannelTypeProvider.java

Author: Do you know if there is something like Number:Percentage for the battery level?

Author: I see that humidity is Number:Dimensionless (and humidity is also %)

| inverterCurrent          | system.electric-current    | The output current of the inverter [A]  |
| inverterVoltage          | system.electric-voltage    | The output voltage of the inverter [V]    |
| inverterFrequency        | Number:ElectricPotential   | The frequency of the output voltage [Hz] |

### Photovoltaic Panels Production Channels

> | Channel                  | Type                       | Description                            |

Reviewer: Can the formatting of this table (in source form) be fixed?

### Battery channels

| Channel                  | Type                       | Description                                                                                    |
|--------------------------|----------------------------|------------------------------------------------------------------------------------------------|
| batteryPower             | system.electric-power      | The power to / from battery (negative means power is pulled from battery and vice-versa) [W]   |
| batteryCurrent           | system.electric-current    | The current to / from battery (negative means power is pulled from battery and vice-versa) [A] |
| batteryVoltage           | system.electric-voltage    | The voltage of the battery [V]                                                                 |
| batteryTemperature       | Number:Temperature         | The temperature of the battery [C/F]                                                           |
| batteryStateOfCharge     | system.battery-level       | The state of charge of the battery [%]                                                         |

### Grid related channels

> | Channel                  | Type                       | Description                                                                                    |

Reviewer: Can the formatting of this table (in source form) be fixed?

Author: I hope I fixed them all...

|--------------------------|----------------------------|------------------------------------------------------------------------------------------------|
| feedInPower              | system.electric-power      | The power to / from grid (negative means power is pulled from the grid and vice-versa) [W]     |
| onGridTotalYield*        | system.electric-power      | Total Yield from PV strings [W]     |
| onGridDailyYield*         | system.electric-power      | The power to / from grid (negative means power is pulled from the grid and vice-versa) [W]     |
| onTotalFeedinEnergy*      | system.electric-power      | The power to / from grid (negative means power is pulled from the grid and vice-versa) [W]     |

> | onTotalFeedinEnergy*      | system.electric-power      | The power to / from grid (negative means power is pulled from the grid and vice-versa) [W]     |

Reviewer: Duplicated.

> _* experimental channels. So far the data does not seem reliable. Either it's taken from wrong part of the JSON or it's calculated wrong in the inverter itself. If someone finds better way to parse it or the meaning of the various int values, please open a new issue for solax binding and share the information._

Reviewer: If they are not working, I would suggest to remove them until it has been figured out how to implement them. You could create the issue already to document your findings and the issue you are having. and track further progress.

### General channels

> | Channel                  | Type                       | Description                                                                                    |

Reviewer: Can the formatting of this table (in source form) be fixed?

Author: Checked and fixed them all to have proper identation of the vertical lines

|--------------------------|----------------------------|------------------------------------------------------------------------------------------|
| lastUpdateTime           | DateTime                   | Last time when a call has been made to the inverter                                      |
| rawData                  | String                     | The raw data retrieved from inverter in JSON format. (Usable for channels not implemented. Can be consumed with the JSONpath transformation |

> ### Properties

Reviewer: It's not strictly needed to document the properties as it's just information.

| Property          | Description                               |
|-------------------|-------------------------------------------|

> | serialWifi        | The wifi module's serial number           |

Reviewer: ```suggestion
| serialWifi        | The Wi-Fi module's serial number          |
```

| inverterType      | Inverter Type (for example X1_HYBRID_G4)  |

## Full Example

> Here are some file based examples

Reviewer: ```suggestion
Here are some file based examples.
```

Number solaxBatteryPower "Battery power [%.0f W]" <energy> (gsolax_inverter,EveryChangePersist) { channel="solax:inverter:localBridge:MyInverter:batteryPower" }
Number solaxBatterySoc "Battery SoC [%.0f %%]" <batterylevel> (gsolax_inverter,EveryChangePersist) { channel="solax:inverter:localBridge:MyInverter:batteryStateOfCharge" }

Number solaxFeedInPower "Feed-in power (CEZ) [%.0f W]" <energy> (gsolax_inverter,EveryChangePersist) { channel="solax:inverter:localBridge:MyInverter:feedInPower" }
Number solaxAcPower "Invertor output power [%.0f W]" <energy> (gsolax_inverter,EveryChangePersist){ channel="solax:inverter:localBridge:MyInverter:inverterOutputPower" }

String solaxInverterType "Inverter Type [%s]" <energy> (gsolax_inverter) { channel="solax:inverter:localBridge:MyInverter:inverterType"}
String solaxUploadTime "Last update time [%s]" <calendar> (gsolax_inverter) { channel="solax:inverter:localBridge:MyInverter:lastUpdateTime" }
String solaxRawData "Raw data [%s]" <data> (gsolax_inverter) { channel="solax:inverter:localBridge:MyInverter:rawData" }
```

### Sitemap Configuration

```perl
Frame label="Solar power strings" {

>                         Text item=solaxPowerEast valuecolor=[<=30="gray",<=300="red", <1500="orange", >=1500="green"] {

Reviewer: Can you reduce the indentation to four space?

Author: wooh... took me a while to manually format this but I did it...

Reviewer: It looks much better now! 👏
